lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/OperatorTopology.cs

```diff
@@ -161,14 +171,18 @@ public void OnNext(GroupCommunicationMessage gcm)
                 throw new ArgumentException("Message must have a source");
             }
 
-            NodeStruct sourceNode = FindNode(gcm.Source);
+            var sourceNode = FindNode(gcm.Source);
             if (sourceNode == null)
             {
                 throw new IllegalStateException("Received message from invalid task id: " + gcm.Source);
             }
 
-            _nodesWithData.Add(sourceNode);
-            sourceNode.AddData(gcm);
+            lock (_thisLock)
```

Comment:
_nodesWithData is already a BlockingCollection, does adding a lock here help anything?